hphp/runtime/vm/jit/vasm-simplify.cpp

```diff
@@ -473,6 +473,29 @@ bool simplify(Env& env, const pop& inst, Vlabel b, size_t i) {
   });
 }
 
+/*
+ * Eliminate masking values with -1 in andXi.
+ * andbi s0=0xff s1=val0 d=val1 -> movb s=val0 d=val1
+ * andli s0=0xffffffff s1=val0 d=val1 -> movl s=val0 d=val1
+ */
+template<typename mov, typename andi>
+bool simplify_andi(Env& env, const andi& inst, Vlabel b, size_t i) {
+  if (inst.s0.l() != -1 ||
+      env.use_counts[andi.sf] != 0) return false;
+  return simplify_impl(env, b, i, [&] (Vout& v) {
+    if (inst.s1 != inst.d) v << mov{inst.s1, inst.d};
```

Comment:
Sorry for the back and forth, but thinking on this a bit more—I think this should actually just be a `copy{}`.  You can nix the template `mov` parameter and just use `copy{}` unconditionally.

The `mov*{}` instructions are only used in unique-stubs, and to be honest, I don't remember right now what their specific semantics are intended to be... on x64, they zero the upper bits of the dst register, but that behavior doesn't seem absolutely required, since currently we only use them to copy constant registers, which are always 64 bits, implicitly.  As far as I can tell, they're basically a `copy{}` that tells the vasm emitter that it's allowed to address less than 64 bits of the operands.

Since `copy{}` is transparent to the register allocator, you don't need the condition, either.  (It's probably slightly better to always emit the `copy{}`, since it avoids a vector shift.)

After that we just need to wait for internal tests to pass before landing.